### PR TITLE
Allow split_reduction if all dyn values are static

### PR DESCRIPTION
**Pull request by voznesenskym**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99483
* __->__ #99475

cc @soumith @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @desertfire

### CODE REVIEW

**Review by ngimel — APPROVED** [review]

File: torch/_inductor/ir.py [hunk 1]
```diff
@@ -825,7 +825,19 @@ def fn(index):
             }
             and config.split_reductions
         )
-        if split_reduction and not dynamo_config.dynamic_shapes:
+        sr_qualified = split_reduction
+        if sr_qualified and dynamo_config.dynamic_shapes:
```
Comment (ezyang) [inline torch/_inductor/ir.py:829]: Is there a reason to gate this on `dynamo_config.dynamic_shapes`? It seems like it should be OK to always do this test

File: torch/_inductor/ir.py [hunk 2]
```diff
@@ -825,7 +825,19 @@ def fn(index):
...
+            # TODO(voz): dedup with sizevar util introduced in other PR
+            def _is_static(x):
+                return isinstance(x, (int, sympy.Integer))
+
+            sr_qualified = (
+                all([_is_static(r) for r in ranges])
+                and all([_is_static(r) for r in reduction_ranges])
+                and _is_static(reduction_numel)
```
Comment (ezyang) [inline torch/_inductor/ir.py:837]: I'll leave @ngimel to check if these conditions make sense

Comment (ezyang) [reply torch/_inductor/ir.py:837]: oops, ngimel approved, okey

File: torch/_inductor/ir.py (hunk 1 above)
Comment (voznesenskym) [reply torch/_inductor/ir.py:829]: Yea, but it's not free, why do a bunch of checks for no reason?

**Comment by voznesenskym** [comment]
Inference TB: 6 % gain, HF: neutral, Timm 1%. Training TB: 2% HF: 1%, Timm 21%

Screenshot 2023-04-19 at 12 50 30 AM
46" alt="Screenshot 2023-04-19 at 12 50 25 AM" src="https://user-images.githubusercontent.com/4755252/233006315-7fd3d32b-ee86-456b-b634-d6adfd212cab.png">

File: torch/_inductor/ir.py (hunk 1 above)
Comment (ezyang) [reply torch/_inductor/ir.py:829]: ...because you're eventually going to turn on dynamic_shapes = True everywhere by default?

**Pull request closed without merging**